thrust/thrust/pair.h

```diff
@@ -88,3 +127,38 @@ using ::cuda::std::make_pair;
  */
 
 THRUST_NAMESPACE_END
+
+_LIBCUDACXX_BEGIN_NAMESPACE_STD
+
+template <class T1, class T2>
+struct tuple_size<THRUST_NS_QUALIFIER::pair<T1, T2>> : tuple_size<pair<T1, T2>>
+{};
+
+template <size_t Id, class T1, class T2>
+struct tuple_element<Id, THRUST_NS_QUALIFIER::pair<T1, T2>> : tuple_element<Id, pair<T1, T2>>
+{};
+
+template <class T1, class T2>
+struct __tuple_like_ext<THRUST_NS_QUALIFIER::pair<T1, T2>> : true_type
+{};
+
+_LIBCUDACXX_END_NAMESPACE_STD
+
+// This is a workaround for the fact that structured bindings require that the specializations of
+// `tuple_size` and `tuple_element` reside in namespace std (https://eel.is/c++draft/dcl.struct.bind#4).
+// See https://github.com/NVIDIA/libcudacxx/issues/316 for a short discussion
+#if _CCCL_STD_VER >= 2017
+
+#  include <utility>
+
+namespace std
+{
+template <class T1, class T2>
+struct tuple_size<THRUST_NS_QUALIFIER::pair<T1, T2>> : tuple_size<pair<T1, T2>>
+{};
+
+template <size_t Id, class T1, class T2>
+struct tuple_element<Id, THRUST_NS_QUALIFIER::pair<T1, T2>> : tuple_element<Id, pair<T1, T2>>
+{};
+} // namespace std
```

Comment:
Matter of style, but saves a few LoCs:
```suggestion
template <class T1, class T2>
struct ::std::tuple_size<THRUST_NS_QUALIFIER::pair<T1, T2>> : std::tuple_size<std::pair<T1, T2>>
{};

template <size_t Id, class T1, class T2>
struct ::std::tuple_element<Id, THRUST_NS_QUALIFIER::pair<T1, T2>> : std::tuple_element<Id, std::pair<T1, T2>>
{};
```

Reply:
Alright, got your point! We also have `tuple_size`, etc. in `::cuda::std`, so it is not automatically clear which one we are specializing.